command/plugins.go

```diff
@@ -53,19 +54,30 @@ func (r *multiVersionProviderResolver) ResolveProviders(
 				continue
 			}
 			if !reqd[name].AcceptsSHA256(digest) {
-				// This generic error message is intended to avoid troubling
-				// users with implementation details. The main useful point
-				// here is that they need to run "terraform init" to
-				// fix this, which is covered by the UI code reporting these
-				// error messages.
-				errs = append(errs, fmt.Errorf("provider.%s: installed but not yet initialized", name))
+				errs = append(errs, fmt.Errorf("provider.%s: checksum mismatch", name))
 				continue
 			}
 
 			client := tfplugin.Client(newest)
 			factories[name] = providerFactory(client)
 		} else {
-			errs = append(errs, fmt.Errorf("provider.%s: no suitable version installed", name))
+			msg := fmt.Sprintf("provider.%s: no suitable version installed", name)
+
+			if req := reqd[name]; req.Versions.String() != "" {
+				foundVersions := []string{}
+				for meta := range r.Available.WithName(name) {
+					foundVersions = append(foundVersions, fmt.Sprintf("%q", meta.Version))
+				}
+
+				found := "none"
+				if len(foundVersions) > 0 {
+					found = strings.Join(foundVersions, ", ")
+				}
+
+				msg += fmt.Sprintf("\n  version requirements: %q\n  versions found: %s", req.Versions, found)
```

Comment:
Thanks, I like that extra paragraph. I was looking through the the other copy for a good example, but was still trying to match the existing init message as closely as possible.